File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/tilgangskontroll/InnloggetBruker.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.tilgangskontroll;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import lombok.Data;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Fnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.altinn.AltinnOrganisasjon;

@Data
public class InnloggetBruker {

  private List<AltinnOrganisasjon> brukerensOrganisasjoner;
  private Fnr fnr;

  public InnloggetBruker(Fnr fnr) {
    this.fnr = fnr;
    brukerensOrganisasjoner = new ArrayList<>();
  }

  public void sjekkTilgang(Orgnr orgnr) {
    if (!harTilgang(orgnr)) {
      throw new TilgangskontrollException("Har ikke tilgang til statistikk for denne bedriften.");
    }
  }

  public boolean harTilgang(Orgnr orgnr) {
    List<String> orgnumreBrukerHarTilgangTil =
        brukerensOrganisasjoner.stream()
            .filter(Objects::nonNull)
            .map(AltinnOrganisasjon::getOrganizationNumber)
            .collect(Collectors.toList());

    return orgnumreBrukerHarTilgangTil.contains(orgnr.getVerdi());
  }
}

File: src/main/kotlin/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domenemodeller/tilgangsstyring/InnloggetBruker.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.tilgangsstyring

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.felles.Orgnr
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.tilgangsstyring.Fnr
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.altinn.AltinnOrganisasjon
import java.util.*
import java.util.stream.Collectors

data class InnloggetBruker(
    val fnr: Fnr,
    var brukerensOrganisasjoner: List<AltinnOrganisasjon> = emptyList()
) {

    fun harTilgang(orgnr: Orgnr): Boolean {
        val orgnumreBrukerHarTilgangTil = brukerensOrganisasjoner.stream()
            .filter { obj: AltinnOrganisasjon? -> Objects.nonNull(obj) }
            .map(AltinnOrganisasjon::organizationNumber)
            .collect(Collectors.toList())
        return orgnumreBrukerHarTilgangTil.contains(orgnr.verdi)
    }
}
